



<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
      
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-4.5.0">
    
    
      
        <title>Other3 - My Wiki Docs</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.1b62728e.css">
      
      
    
    
      <script src="../assets/javascripts/modernizr.268332fc.js"></script>
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../assets/fonts/material-icons.css">
    
    
    
      
    
    
  </head>
  
    <body dir="ltr">
  
    <svg class="md-svg">
      <defs>
        
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#macs2p-valuepeak10000" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href=".." title="My Wiki Docs" class="md-header-nav__button md-logo">
          
            <i class="md-icon"></i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            <span class="md-header-nav__topic">
              My Wiki Docs
            </span>
            <span class="md-header-nav__topic">
              
                Other3
              
            </span>
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
          
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
        
      </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main" role="main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href=".." title="My Wiki Docs" class="md-nav__button md-logo">
      
        <i class="md-icon"></i>
      
    </a>
    My Wiki Docs
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../ASM/" title="First" class="md-nav__link">
      First
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../clusterProfiler/" title="Second" class="md-nav__link">
      Second
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../CpG_island_classification/" title="Third" class="md-nav__link">
      Third
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../methy_Haplo/" title="Other1" class="md-nav__link">
      Other1
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../PMD/" title="Other2" class="md-nav__link">
      Other2
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
    <a href="./" title="Other3" class="md-nav__link md-nav__link--active">
      Other3
    </a>
    
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                
                <p><strong>Binding and Expression Target Analysis (BETA)</strong></p>
<p>该分析整合了ChIP-seq和转录表达水平来探究基因表达调控的机理。此前，一些研究用于转录因子靶基因预测，但是鲜有高效的工具供人们使用。</p>
<p>作者开发了名为BETA的软件来研究结合位点与转录表达的关系，基于结合位点信息和差异表达信息可以进行如下三个主要分析：</p>
<p>​   （1）用于预测转录因子就有激活或抑制的功能；</p>
<p>​   （2）用于推断转录因子的靶基因；</p>
<p>​   （3）用于鉴定转录因子的motif及其结合者。</p>
<p>该软件具有三个主要的子命令：</p>
<p><strong>第一部分：</strong></p>
<p>BETA basic： 转录因子激活或抑制功能预测和直接靶基因检测。</p>
<pre><code class="shell">BETA basic \
    –p peaks.bed \
    –e expr.xls  \
    –k BSF  \
    –g hg19 \
    --da 500 \
    –n prefix \
    -o outdir 
</code></pre>

<p>涉及参数及其他 可选参数解释如下：</p>
<p>-p  peaks结合位点文件，BED格式，支持3列或5列格式，其格式为：</p>
<p>​   CHROM, START, END（NAME, SCORE）。</p>
<p>使用基本的3列信息即可。</p>
<p>示例如下（不包含表头）：</p>
<table>
<thead>
<tr>
<th align="left">#chrom</th>
<th>start</th>
<th>end</th>
<th>name（可选）</th>
<th>score（可选）</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">chr1</td>
<td>1208689</td>
<td>1209509</td>
<td>AR_LNCaP_2</td>
<td>51.58</td>
</tr>
<tr>
<td align="left">chr1</td>
<td>1334246</td>
<td>1335348</td>
<td>AR_LNCaP_7</td>
<td>54.55</td>
</tr>
<tr>
<td align="left">chr1</td>
<td>2179351</td>
<td>2180790</td>
<td>AR_LNCaP_9</td>
<td>257.72</td>
</tr>
</tbody>
</table>
<h3 id="macs2p-valuepeak10000">macs2检峰时刻通过p-value进行peak的过滤只保留10000个结果。</h3>
<p>-e  差异表达文件，可以使用LIMMA和Cuffdiff的结果，也可使用BETA specific format（BSF）格式，其格式如下：</p>
<table>
<thead>
<tr>
<th>GeneID</th>
<th>log2(FC)</th>
<th>FDR/pval</th>
</tr>
</thead>
<tbody>
<tr>
<td>NM_000015</td>
<td>1.28221</td>
<td>0.185744</td>
</tr>
<tr>
<td>NM_000016</td>
<td>0</td>
<td>1</td>
</tr>
<tr>
<td>NM_000017</td>
<td>0.0598207</td>
<td>0.844755</td>
</tr>
</tbody>
</table>
<p>-k  差异表达文件的类型，即LIM, CUF, BSF, O，分别对应LIMMA、cuffdiff标准输出格式、BSF格式或其他格式的文件需要使用--info指定。</p>
<p>-g  基因组，支持任何小鼠，hg38, hg19, hg18, mm10, mm9，也可使用-r指定其他基因组文件。</p>
<p>--gname2    给定次参数说明文件中的基因名称或转录本名称是官方的基因symbols，默认FALSE。</p>
<p>--info  指定表达数据中的geneID, up/down status和statistcal values列。</p>
<p>-r  参考基因组文件。</p>
<p>-o  输出路径。</p>
<p>--bl    是否使用CTCF边界过滤基因周围的peaks，默认FALSE。</p>
<p>--bf    CTCF保守peakBED文件，需要同时指定--bl参数且基因组不能指定为hg19和mm9。</p>
<p>--pn    打算分析的peaks数目，默认10000。</p>
<p>--method    支持score和distance，指定TF/CR给你预测的方法，score：调控潜能的打分，disrance：邻近的结合peak的距离。</p>
<p>-n  输出文件前缀名称。</p>
<p>-d  从TSS一定范围内获取peaks，默认100000（100kb）。</p>
<p>--df    输入一个0~1范围内的阈值来筛选最显著的差异表达基因，默认1，全部输入基因。</p>
<p>--da    通过比例或数目选取最显著的差异表达基因，默认0.5。如果想使用diff_fdr请设置此参数为1。</p>
<p>-c  设置0~1阈值，通过p值（单尾KS检验）选择最近的靶基因，默认0.001。</p>
<p>测试运行日志如下：</p>
<pre><code>BETA basic \
  -p ../peaks.bed \
  -e ../beta_specific_expression.txt \
  -k BSF \
  -g hg19 \
  --da 500 \
  -n basic \
  -o basic_out

[15:41:26] Argument List:
[15:41:26] Name = basic
[15:41:26] Peak File = ../peaks.bed
[15:41:26] Top Peaks Number = 10000
[15:41:26] Distance = 100000 bp
[15:41:26] Genome = hg19
[15:41:26] Expression File = ../beta_specific_expression.txt
[15:41:26] BETA specific Expression Type
[15:41:26] Number of differential expressed genes = 500.0
[15:41:26] Differential expressed gene FDR Threshold = 1
[15:41:26] Up/Down Prediction Cutoff = 0.001000
[15:41:26] Function prediction based on regulatory potential
[15:41:26] Check ../peaks.bed successfully!
[15:41:26] #ID  Status  Value
 is the header of the expression file
[15:41:26] Checking the differential expression infomation...
[15:41:26] Take the first line with Differential Information as an example: NM_000014   -0.325878       0.618293

[15:41:26] Differential Expression file format successful passed
[15:41:26] You do not like filter peak by CFCT boundary, it will be filtered only by the distance
[15:41:26] Read file &lt;../peaks.bed&gt; OK! All &lt;7031&gt; peaks.
[15:41:41] Process &lt;50801&gt; genes
[15:41:56] Finished! Preliminary results saved into temporary file: &lt;basic.txt&gt;
[11815, 11643]
[15:41:56] Genes were seprated to two parts: up regulated and down regulated.
cut: write error: Broken pipe
cut: write error: Broken pipe
[15:41:56] Prepare file for the Up/Down Test
null device
          1
[15:42:02] Finished, Find the result in basic_function_prediction.pdf
[15:42:02] Get the Rank Product of the &quot;upregulate&quot; genes
[15:42:46] Get the Rank Product of the &quot;downregulate&quot; genes
['&quot;upregulate&quot;', '&quot;downregulate&quot;']
[15:43:30] pick out the peaks 100000bp around the selected genes
[15:43:30] Finished: Find target gene associated peaks in basic_out
total time: 0:2:4
</code></pre>

<p>输出文件存储于basic_out文件夹：</p>
<pre><code class="shell">$tree
.
├── basic_downtarget_associate_peaks.bed
├── basic_downtarget.txt
├── basic_function_prediction.pdf
├── basic_function_prediction.R
├── basic_uptarget_associate_peaks.bed
└── basic_uptarget.txt
</code></pre>

<ol>
<li>激活/抑制功能预测打分</li>
</ol>
<p><img alt="mkdocs" src="../images/beta_1.png" /></p>
<ol>
<li>上调调控靶标/下调调控靶标文件（basic_uptarget.txt，basic_downtarget.txt）</li>
</ol>
<pre><code>#文件截取示例
Chroms  txStart txEnd   refseqID        rank product    Strands GeneSymbol
chr1    27992571        27998724        NM_002038       4.592e-06       -       IFI6
chr12   113376237       113411055       NM_006187       4.835e-06       +       OAS3
chr12   113344738       113357712       NM_016816       5.523e-06       +       OAS1
chr4    169277885       169401665       NM_001012967    5.616e-06       -       DDX60L
</code></pre>

<h1 id="_1">以基因为中心</h1>
<ol>
<li>靶基因相关的peak文件（是以peak为中心）</li>
</ol>
<pre><code>chrom   pStart  pEnd    Refseq  Symbol  Distance        Score
chr1    27971570        27971739        NM_002038       IFI6    -27070  0.205399174599
chr1    27971776        27971885        NM_002038       IFI6    -26894  0.20685028671
chr12   113364362       113364438       NM_006187       OAS3    -11837  0.377766121894
chr12   113364362       113364438       NM_016816       OAS1    19662   0.276241443606
chr4    169422368       169422769       NM_001012967    DDX60L  20903   0.262863603291
</code></pre>

<h1 id="peak">以peak为中心</h1>
<p><strong>第二部分：</strong></p>
<p>BETA plus：进行激活/抑制功能预测， 直接靶基因预测和靶向区域motif分析，比基础版本增加了motif分析</p>
<pre><code>BETA plus \
    –p peaks.bed \
    –e expr.xls  \
    –k BSF \
    –g hg19 \
    --gs hg19.fa \
    --bl
</code></pre>

<p>该模块与basic相似，以下参数功能有差异：</p>
<p>--gs    基因序列文件。</p>
<p>--mn    指定0~1范围内数值作为p-value或大于1作为数目来获取motif的阈值， 默认10。</p>
<p>其他参数参考<strong>BETA Basic</strong>参数描述。</p>
<p>测试运行日志如下：</p>
<pre><code>[16:49:30] Argument List:
[16:49:30] Name = NA
[16:49:30] Peak File = ../peaks.bed
[16:49:30] Top Peaks Number = 10000
[16:49:30] Distance = 100000 bp
[16:49:30] Genome = hg19
[16:49:30] Expression File = ../beta_specific_expression.txt
[16:49:30] Genome Sequence fasta formated data = hg19.fa
[16:49:30] BETA specific Expression Type
[16:49:30] Number of differential expressed genes = 0.5
[16:49:30] Differential expressed gene FDR Threshold = 1
[16:49:30] Up/Down Prediction Cutoff = 0.001000
[16:49:30] Function prediction based on regulatory potential
[16:49:30] Check ../peaks.bed successfully!
[16:49:30] #ID  Status  Value
 is the header of the expression file
[16:49:30] Checking the differential expression infomation...
[16:49:30] Take the first line with Differential Information as an example: NM_000014   -0.325878       0.618293

[16:49:30] Differential Expression file format successful passed
[16:49:30] Read file &lt;../peaks.bed&gt; OK! All &lt;7031&gt; peaks.
[16:50:16] Process &lt;50801&gt; genes
[16:50:25] Finished! Preliminary results saved into temporary file: &lt;NA.txt&gt;
[11815, 11643]
[16:50:25] Genes were seprated to two parts: up regulated and down regulated.
cut: write error: Broken pipe
cut: write error: Broken pipe
[16:50:26] Prepare file for the Up/Down Test
null device
          1
[16:51:03] Finished, Find the result in NA_function_prediction.pdf
[16:51:03] Get the Rank Product of the &quot;upregulate&quot; genes
[16:51:45] Get the Rank Product of the &quot;downregulate&quot; genes
['&quot;upregulate&quot;', '&quot;downregulate&quot;']
[16:52:26] pick out the peaks 100000bp around the selected genes
[16:52:27] Finished: Find target gene associated peaks in plus_out
[16:52:27] get three regions of every peak around the up genes
['up_middle.bed', 'up_left.bed', 'up_right.bed']
[16:53:54] get the fasta format sequence data of the three regions of up
[16:53:55] get three regions of every peak around the down genes
['down_middle.bed', 'down_left.bed', 'down_right.bed']
[16:55:14] get the fasta format sequence data of the three regions of down
[16:55:14] get three regions of every peak around the non genes
['non_middle.bed', 'non_left.bed', 'non_right.bed']
[16:56:37] get the fasta format sequence data of the three regions of non
[16:56:37] run mis to do the known motif scan with cistrome motif database
Calculating Background..
A or T: 93234
C or G: 100766
Others: 0
Calculating Background..
A or T: 97696
C or G: 95334
Others: 0
Calculating Background..
A or T: 97264
C or G: 95766
Others: 0
Calculating Background..
A or T: 118812
C or G: 142588
Others: 0
Calculating Background..
A or T: 125071
C or G: 135022
Others: 0
Calculating Background..
A or T: 123892
C or G: 136201
Others: 0
Calculating Background..
A or T: 181589
C or G: 169211
Others: 0
Calculating Background..
A or T: 191633
C or G: 157413
Others: 0
Calculating Background..
A or T: 191487
C or G: 157559
Others: 0
[16:59:58] T statistic test performed here to do the significance testing of every motif
[17:58:18] motif result will be in html format
Success parser from table.
Success parser from table.
Success parser from table.
Success parser from table.
Success parser from table.
Success parser from table.
[17:58:22] Done: find motif result in beatmotif.html file
[17:58:22] Done: find all BETA result in plus_out
total time: 1:8:52
</code></pre>

<p>同样，plus功能生成上/下调靶标文件和靶基因相关的peak文件，此外生成了靶基因相关motif的结果文件（存放于motifresult文件夹，包含一个*motif.html的motif分析文件）。</p>
<p>motif分析示例截图如下：</p>
<p><img alt="mkdocs" src="../images/beta_2.png" /></p>
<p>同时生成上调、下调差异差异motif的文件：</p>
<p><strong>MotifID   Species Symbol  DNA BindDom PSSM    Tscore  Pvalue</strong></p>
<p>MS00266 Homo sapiens    MLX Helix-Loop-Helix Family [[[0.534, 0.012, 0.331, 0.123], [0.01, 0.043, 0.152, 0.795], [0.011, 0.963, 0.011, 0.015], [0.955, 0.01, 0.025, 0.01], [0.01, 0.97, 0.01, 0.01], [0.01, 0.01, 0.97, 0.01], [0.01, 0.01, 0.01, 0.97], [0.01, 0.01, 0.97, 0.01], [0.8, 0.03, 0.117, 0.053], [0.066, 0.241, 0.017, 0.676]]]    5.08    4.30e-07</p>
<h2 id="pssmposition-specific-scoring-matrixpssm-020303020108005005-tscoret-test">PSSM：位置特异性得分矩阵（position-specific scoring matrix），PSSM示例格式为: [[0.2,0.3,0.3,0.2],[0.1,0.8,0.05,0.05]]； Tscore：t-test的统计量数值。</h2>
<p>另外，官网提供了motif分析结果的展示示例参考：</p>
<p>​    http://cistrome.org/BETA/motifresult/betamotif.html。</p>
<p><strong>第三部分：</strong></p>
<p>用于预测TF的靶基因和激活/抑制功能，属于basic的简化模块。</p>
<pre><code class="shell">BETA minus \
    -p peaks.bed \
    --bl -g hg19 
</code></pre>

<h1 id="basic">参数可参考basic参数解释。</h1>
<p><strong>参考资料</strong></p>
<p>官方使用手册：http://cistrome.org/BETA/tutorial.html</p>
<p>BETA网页版分析入口：http://cistrome.org/BETA/index.html#runweb</p>
<p>转录因子motif知识：https://www.jianshu.com/p/ad82fe903e57</p>
<p><strong>参考文献</strong></p>
<p>Wang, S., Sun, H., Ma, J., Zang, C., Wang, C., Wang, J., ... &amp; Liu, X. S. (2013). Target analysis by integration of transcriptome and ChIP-seq data with BETA. <strong>Nature protocols</strong>, 8(12), 2502-2515.</p>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../PMD/" title="Other2" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Other2
              </span>
            </div>
          </a>
        
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.a8b5e56f.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:".."}})</script>
      
    
  </body>
</html>